app/protocol/flows/blockrelay/ibd.go

```diff
@@ -118,40 +118,57 @@ func (flow *handleRelayInvsFlow) syncHeaders(highHash *externalapi.DomainHash) e
 	return nil
 }
 
-func (flow *handleRelayInvsFlow) findHighestSharedBlockHash(highHash *externalapi.DomainHash) (
-	lowHash *externalapi.DomainHash, err error) {
-
-	lowHash = flow.Config().ActiveNetParams.GenesisHash
-	currentHighHash := highHash
+func (flow *handleRelayInvsFlow) findHighestSharedBlockHash(targetHash *externalapi.DomainHash) (*externalapi.DomainHash, error) {
+	lowHash := flow.Config().ActiveNetParams.GenesisHash
+	highHash, err := flow.Domain().Consensus().GetHeadersSelectedTip()
+	if err != nil {
+		return nil, err
+	}
 
-	for {
-		err := flow.sendGetBlockLocator(lowHash, currentHighHash, 0)
+	for !lowHash.Equal(highHash) {
+		log.Debugf("Sending a blockLocator to %s between %s and %s", flow.peer, lowHash, highHash)
+		blockLocator, err := flow.Domain().Consensus().CreateBlockLocator(lowHash, highHash, 0)
 		if err != nil {
 			return nil, err
 		}
 
-		blockLocatorHashes, err := flow.receiveBlockLocator()
+		ibdBlockLocatorMessage := appmessage.NewMsgIBDBlockLocator(targetHash, blockLocator)
+		err = flow.outgoingRoute.Enqueue(ibdBlockLocatorMessage)
 		if err != nil {
 			return nil, err
 		}
-
-		// We check whether the locator's highest hash is in the local DAG.
-		// If it is, return it. If it isn't, we need to narrow our
-		// getBlockLocator request and try again.
-		locatorHighHash := blockLocatorHashes[0]
-		locatorHighHashInfo, err := flow.Domain().Consensus().GetBlockInfo(locatorHighHash)
+		message, err := flow.dequeueIncomingMessageAndSkipInvs(common.DefaultTimeout)
 		if err != nil {
 			return nil, err
 		}
-		if locatorHighHashInfo.Exists {
-			return locatorHighHash, nil
+		ibdBlockLocatorHighestHashMessage := message.(*appmessage.MsgIBDBlockLocatorHighestHash)
```

Comment:
The incoming route for this flow registers to many message types. If the message received is not MsgIBDBlockLocatorHighestHash - we'll crash.
Instead, do a checked type assertion, and return a protocol error if it's not ok.